Lightweight importer

dominikl opened the pull request:
# What this PR does

If more than 100 images are selected for import the UI doesn't show a detailed UI component (with progress bars, buttons, etc.) for each file any longer, but just displays a simplified UI to track the overall import process.

![screen shot 2018-06-15 at 14 12 12](https://user-images.githubusercontent.com/6575139/41470312-28a0ba7e-70a8-11e8-87e0-0fcab06ded0f.png)

# Testing this PR

See also [Import workflow scenario](https://docs.openmicroscopy.org/internal/testing_scenarios/ImportUI.html)

Check that the general import workflow still works, with "import batches" less than 100 files.

Check the import workflow also works with the simplified UI, i. e. for "import batches" larger than 100 files. I tested it with 3000 files,  worked fine. There are still tons of objects created (non-UI at least), but the heap usage seems to stay under 500 MB and the import process didn't slow down (as it does with the "heavy UI" until it freezes completely). 

Include one or more broken files and check that these cases are handled correctly, i.e. show up as "failed", you can get the full error log and it doesn't "block" the importer (you don't have to "force quit" it).

Check the new "skip" options.

![screen shot 2018-07-03 at 16 09 06](https://user-images.githubusercontent.com/6575139/42228385-bf817fb4-7edb-11e8-86ac-97600fee9ee2.png)


# Related reading

https://docs.google.com/document/d/1e3jeIlYxzI9w5hEdD008gvIH6Q70xig7nplzmaaIpw8/edit#heading=h.bee3vpgcgqg2

dominikl committed:
Separate UI code from StatusLabel

dominikl committed:
Add UI-less ImporterUIElement and FileImportComponent

dominikl committed:
Properly separate ImporterUIElement

dominikl committed:
Fixed PropertyChangeEvents and improvied UI

rgozim committed:
Added advanced panel under projects/screen tab (show/hide not working)

rgozim committed:
Reverted LocationDialog
Added options for skipping import work to ImportDialog

rgozim committed:
Improved tooltips for checkboxes (matches docs)
Added data placeholder in ImportableObject

dominikl committed:
Some cleanup

dominikl committed:
Pass on skip options to ImportConfig / ImportCandidates

dominikl commented:
@rgozim PR is included now. I also added the code to pass the skip options on to the ImportConfig / ImportCandidates, seems to work (at least for the skip thumbnails, not verified the other options yet). 

It looks like I added a bug with the respect to the lightweight import. Import errors aren't registered any longer and the spinner just keeps spinning in that case. Will have to debug/fix that, so not ready for review yet unfortunately.

dominikl committed:
Fixed some javadoc issues

dominikl committed:
Show import errors

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/ImportDialog.java:
```diff
@@ -26,12 +26,7 @@
-import java.awt.Font;
-import java.awt.GridBagConstraints;
-import java.awt.GridBagLayout;
+import java.awt.*;
```
We tend to avoid wildcard imports.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/FailedImportDialog.java:
```diff
@@ -0,0 +1,168 @@
+        con.add(new JLabel("File:"), c);
+
+        String[] tmp = null;
+        if (errors.keySet().size() < 50) {
```
Maybe replace the 50-related stuff into something using a `final int WHATEVER = 50`?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUI.java:
```diff
@@ -66,6 +67,9 @@
 import javax.swing.text.StyledDocument;
 
+
+
```
rather than insert these, fine to get rid of the `//` comments

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUI.java:
```diff
@@ -79,10 +83,13 @@
 import org.openmicroscopy.shoola.env.data.model.ImportableFile;
 import org.openmicroscopy.shoola.env.data.model.ImportableObject;
+
 import omero.gateway.SecurityContext;
```
rather than insert these blanks, might just make sense to move this line

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/model/ImportableObject.java:
```diff
@@ -263,6 +278,20 @@ public ImportableObject(List<ImportableFile> files, boolean overrideName)
+     * @param choices
+     *            The options
+     */
+    public void setSkipChioces(Map<String, Object> choices) {
```
setSkipChoices

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/model/ImportableObject.java:
```diff
@@ -648,4 +679,44 @@ public void reUpload(List<ImportableFile> files)
+     */
+    public String getChecksumAlgorithm() {
+        String option = (String) skipChoices.get(OPTION_CHECKSUMS);
+        return option != null ? option : "File-Size-64";
```
Probably good to use the constants themselves where possible, e.g., `omero.model.enums.ChecksumAlgorithmFileSize64.value`.

dominikl committed:
Fixed imports; formatting; use omero.model.enums

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/ImportDialog.java:
```diff
@@ -1095,7 +1108,7 @@ private int handleFilesSelection(File[] files) {
 
 	/** Imports the selected files. */
-	public void importFiles() {
+	public void  importFiles() {
```
probably not intended

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/SkipComputePanel.java:
```diff
@@ -0,0 +1,140 @@
+ */
+package org.openmicroscopy.shoola.agents.fsimporter.chooser;
+
+import javax.swing.*;
```
avoid wildcard imports

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/SkipComputePanel.java:
```diff
@@ -0,0 +1,140 @@
+            @Override
+            public void itemStateChanged(ItemEvent e) {
+                if (checksumCheckBox.isSelected()) {
+                    choices.put(ImportableObject.OPTION_CHECKSUMS, "File-Size-64"); // File-Size-64 fast
```
can use `omero.model.enums.ChecksumAlgorithmFileSize64.value` etc.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponentI.java:
```diff
@@ -0,0 +1,404 @@
+public interface FileImportComponentI {
+
+    /** Indicates that the container is of type <code>Project</code>. */
+    public static final int PROJECT_TYPE = 0;
```
Could a new inner `enum` be appropriate?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponentI.java:
```diff
@@ -0,0 +1,404 @@
+    public abstract FileObject getFile();
+
+    /**
+     * Returns the file hosted by this component.
```
Maybe a clue as to why to choose this instead of the above `getFile()`?

dominikl replied:
No idea. There's a 'file' and an 'originalFile' in the ImportableFile, where the originalFile apparently is "The file object before possible changes". No clue, what that's supposed to mean.

mtbc replied:
Heh, understood, It probably made sense to somebody long ago.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponentI.java:
```diff
@@ -0,0 +1,404 @@
+    public abstract boolean hasFailuresToReimport();
+
+    /**
+     * Returns <code>true</code> if the file can be re-imported,
```
very similar to `hasFailuresToReimport()`?

dominikl replied:
That's true, both methods have the same implementation. I remove one of them.

dominikl replied:
Looks like I was wrong, I have to leave both methods in.

mtbc replied:
Maybe Javadoc can explain the difference?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponentI.java:
```diff
@@ -0,0 +1,404 @@
+    public abstract boolean hasFailuresToReupload();
+
+    /**
+     * Returns <code>true</code> if the import has failed, <code>false</code>
```
or maybe if it started?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponentI.java:
```diff
@@ -0,0 +1,404 @@
+    public abstract boolean hasComponents();
+
+    /**
+     * Returns the status of the import process one of the
```
maybe some punctuation before "one"?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1154 @@
+            containerObject = dataset;
+            return;
+        }
+        if (data != null && data instanceof ScreenData) {
```
No need for the null check.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1154 @@
+     */
+    private FileObject getAssociatedFile(int series) {
+        List<FileObject> l = getFile().getAssociatedFiles();
+        Iterator<FileObject> i = l.iterator();
```
seems to be a long way to write,
```java
for (FileObject f : getFile().getAssociatedFiles()) {
    if (f.getIndex() == series) {
    ...
```
Fine to use Java 5 in new code. (Though nothing uses this method anyway.)

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1154 @@
+        if (importable.getGroup() != null)
+            buf.append("_" + importable.getGroup().getId());
+        if (importable.getUser() != null)
+            buf.append("_" + importable.getUser().getId());
```
If there's either a group or a user we can't tell which kind of ID it is?

dominikl replied:
I remove the toString methods, guess they were used for debug purposes once.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterComponent.java:
```diff
@@ -50,7 +51,9 @@
 import org.openmicroscopy.shoola.env.data.model.ResultsObject;
 import org.openmicroscopy.shoola.env.data.model.ThumbnailData;
+
 import omero.gateway.SecurityContext;
```
rather than inserting lines maybe just move this to its own section

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/util/Status.java:
```diff
@@ -0,0 +1,681 @@
+    private Object callback;
+
+    /** Indicates that the file scanned is a directory. */
+    // private boolean directory;
```
may wish to omit altogether

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/util/Status.java:
```diff
@@ -0,0 +1,681 @@
+        firePropertyChange(UPLOAD_DONE_PROPERTY, null, this);
+    }
+
+    /** Marks the import has cancelled. */
```
maybe s/has/as/

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/util/Status.java:
```diff
@@ -0,0 +1,681 @@
+        return markedAsCancel;
+    }
+
+    /** Marks the import has duplicate. */
```
maybe s/has/as/

dominikl committed:
Use enum

dominikl committed:
Add simpler UI

dominikl committed:
Some cleanup

dominikl committed:
More clean up

dominikl commented:
Made the UI a bit simpler, guess two progress bars and an error count is enough. If there is an error, the button "Show failed" at the top becomes active and provides error details.

![screen shot 2018-07-13 at 15 49 42](https://user-images.githubusercontent.com/6575139/42698619-56339c1c-86b6-11e8-9c15-5250770af2eb.png)

![screen shot 2018-07-13 at 15 55 08](https://user-images.githubusercontent.com/6575139/42698608-4e4b12f0-86b6-11e8-8231-970bff4f31fd.png)

kennethgillen commented:
@dominikl - I'm no HCI expert, but the text under the complete bar is pretty hard to read.

![lightweight_importer_by_dominikl_ _pull_request__5783_ _openmicroscopy_openmicroscopy](https://user-images.githubusercontent.com/2727871/42698841-f9d74d8c-86b6-11e8-8a64-e0a56cecd655.png)

Is it possible to add some kind of blur, mask or effect behind the text, to make it readable? e.g. https://www.slideteam.net/blog/11-hacks-to-make-text-over-images-more-readable-craft-a-stunning-slide/

dominikl commented:
Quite tricky change the L&F of the progressbar. An easier alternative would be to move the text out of the progress bar itself:
![screen shot 2018-07-13 at 16 37 53](https://user-images.githubusercontent.com/6575139/42700275-24570c10-86bb-11e8-856a-70edd06b1d75.png)

kennethgillen commented:
Nice @dominikl - looks far more readable to me. Let's see what others think.

dominikl committed:
Improved UI a bit

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/ROISaver.java:
```diff
@@ -24,9 +24,11 @@
+import org.openmicroscopy.shoola.agents.fsimporter.util.FileImportComponentI;
 import org.openmicroscopy.shoola.agents.fsimporter.view.Importer;
+
 import omero.gateway.SecurityContext;
```
move this rather than introducing the blank lines

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/ImportDialog.java:
```diff
@@ -950,9 +932,17 @@ private JComponent buildNamingComponent() {
+	private JComponent buildSkipFlagsComponent() {
+		skipComputePanel = new SkipComputePanel();
+		JPanel content = new JPanel();
+		content.setBorder(BorderFactory.createTitledBorder("Skip Compute"));
```
Intentional not to reuse `TEXT_SKIP_COMPUTE`?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/ImportDialog.java:
```diff
@@ -1726,3 +1742,4 @@ public void actionPerformed(ActionEvent evt) {
 	}
 
 }
+
```
guessing unintentional

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/FileImportComponent.java:
```diff
@@ -77,7 +76,10 @@
 import org.openmicroscopy.shoola.env.data.model.ImportableFile;
 import org.openmicroscopy.shoola.env.data.model.ThumbnailData;
+
 import omero.gateway.SecurityContext;
```
Move this rather than introducing spaces around it.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+    /** Indicates the status of the on-going import. */
+    private Status status;
+
+    /** Keep tracks of the components. */
```
track

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+        this.data = data;
+        this.dataset = dataset;
+        this.refNode = refNode;
+        if (refNode != null && refNode instanceof TreeImageDisplay) {
```
don't need the null check

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+                    while (i.hasNext()) {
+                        fc = i.next();
+                        list = fc.getImportErrors();
+                        if (!CollectionUtils.isEmpty(list))
```
FYI there's an `isNotEmpty`

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+            Iterator<LightFileImportComponent> i = values.iterator();
+            while (i.hasNext()) {
+                if (i.next().hasImportStarted())
+                    count++;
```
It feels as if this can just be,
```java
synchronized (components) {
    for (LightFileImportComponent component : components.values()) {
        if (!component.hasImportStarted()) {
            return false;
        }
    }
}
return true;
```
It's a curious method though: import hasn't started if any of its components haven't?

dominikl replied:
Indeed. I guess it's one of these methods which aren't actually doing what one would expect from its name `hasImportStarted`. I only tried to separate the UI code from the functionality, without looking into these methods with much detail. But I'll simply the method like you suggested.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+    private Object refNode;
+
+    /** The object where the data have been imported. */
+    private DataObject containerObject;
```
I can't quite tell what we actually do with this.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/util/LightFileImportComponent.java:
```diff
@@ -0,0 +1,1135 @@
+     */
+    @Override
+    public boolean toRefresh() {
+        /*
```
Guessing this is work in progress.

dominikl replied:
If so it has been work in progress since a long time, as far as I can see this always has been commented out.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/FailedImportDialog.java:
```diff
@@ -0,0 +1,171 @@
+public class FailedImportDialog extends JDialog {
+
+    /** Holds the error logs (key: file name) */
+    private Map<String, String> errors = new HashMap<String, String>();
```
Would it be nice to make this a `SortedMap` for presentation?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterComponent.java:
```diff
@@ -50,12 +51,13 @@
 import org.openmicroscopy.shoola.env.data.model.ResultsObject;
 import org.openmicroscopy.shoola.env.data.model.ThumbnailData;
-import omero.gateway.SecurityContext;
+
```
odd choice for a blank line

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterControl.java:
```diff
@@ -58,13 +58,14 @@
-import org.openmicroscopy.shoola.env.data.util.StatusLabel;
-import omero.log.Logger;
+import org.openmicroscopy.shoola.env.data.util.Status;
+
```
odd blank line

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUIElement.java:
```diff
@@ -69,1155 +70,1023 @@
+    {
+        StringBuffer buffer = new StringBuffer();
+        int n = countUploaded-countUploadFailure-countCancelled;
+        if (n < 0) n = 0;
```
I wonder if this actually happens?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUIElementDetailed.java:
```diff
@@ -0,0 +1,312 @@
+	private static final double[] COLUMNS = {TableLayout.FILL};
+	
+	/** The message to display in the header.*/
+	private static final String MESSAGE = 
```
Maybe we should try to reuse what we can from `ImporterUIElement`.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUIElementLight.java:
```diff
@@ -0,0 +1,262 @@
+
+        if (super.countFailure > 0) {
+            super.filterButton.setEnabled(true);
+        }    }
```
separate lines for these

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/model/ImportableObject.java:
```diff
@@ -431,6 +460,8 @@ public Class getRootType()
 	}
+
+
 	
```
unintentional?

mtbc commented:
It's a pity this isn't easier to write tests for: I have a fear that when we overhaul import, even just adding another intermediate step, some of the step-conscious code will break. Definitely worth looking for opportunities to localize anything that checks steps numbers so we can someday easily find what needs fixing.

dominikl committed:
Some code clean up

dominikl committed:
Removed unused imports

dominikl commented:
Just a note: Still have to check/fix this issue: https://www.openmicroscopy.org/community/viewtopic.php?f=4&t=8560 in this PR as well.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/SkipComputePanel.java:
```diff
@@ -0,0 +1,142 @@
+     * Text for checkboxes
+     */
+    private static final String TEXT_THUMBNAIL_GENERATION = "Thumbnail generation";
+    private static final String TEXT_MIN_MAX_COMPUTE = "Min/max compute";
```
I think we've usually said "calculate" for min/max?

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/ImportDialog.java:
```diff
@@ -177,7 +178,13 @@
+	private static final String TEXT_FILE_NAMING = "File Naming";
+
+	/** Text for metadata pane */
+	private static final String TEXT_SKIP_COMPUTE = "Skip Compute";
```
Maybe "Skip Processing"? "Compute" may be new to OMERO UIs as a noun.

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/chooser/SkipComputePanel.java:
```diff
@@ -0,0 +1,142 @@
+     */
+    private static final String TEXT_THUMBNAIL_GENERATION = "Thumbnail generation";
+    private static final String TEXT_MIN_MAX_COMPUTE = "Min/max compute";
+    private static final String TEXT_CHECKSUM_CHECKING = "Checksum checking";
```
maybe "Checksum verification"

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUIElementLight.java:
```diff
@@ -0,0 +1,263 @@
+        c.gridx = 0;
+        c.gridy = 3;
+        c.fill = GridBagConstraints.NONE;
+        info.add(new JLabel("Errors:"), c);
```
Would be nice to right-justify these labels.

mtbc commented:
Definitely running into some "force quit" scenarios. The most obvious so far: from `ome/data_repo/test_images_good/dv/` select both `CFPNEAT01_R3D.dv` and `CFPNEAT01_R3D.dv.log` then try to import.

mtbc commented:
After bulk import it would be nice to have some reassurance that checksums did verify okay.

dominikl commented:
Ok, after my recent changes I broke the importer again (only imports the first file, then stops). I'll revert and really only fix trivial things. Looks like any change to existing methods has a high risk of unexpected side effects.

dominikl committed:
Used SortedMap for failed import dialog

dominikl committed:
Bring back the essential(!) toString method

dominikl commented:
Actually, https://www.openmicroscopy.org/community/viewtopic.php?f=4&t=8560 is already fixed with this PR (/cc @joshmoore ). I'll go once more over all comments and see what I can address and then list it for testing on Monday.

joshmoore reviewed (commented):
*Lightweight GUI* (@dominikl):
 * 100+ uploads are definitely simplified.
 * I wonder if we need to explain why the UI is different, either visually or just with text about "Uploaded: Processed:".

*Skip compute* (@rgozim):
 * I don't think that UpgradeCheck is needed since unfortunately (for code clarify) that option is only used in CommandLineImporter. Insight will have already performed a single upgrade check on startup.
 * This may be a larger issue with the state of the import dialog, but the options aren't reset between imports. (e.g. I also noted that new contents of a directory on Mac don't show up until I navigate to another directory and come back.)
 * I assume it's too complicated for the current UI, but perhaps something to think about when designing the next version is to include the "skip compute" options in the summary display.

I will look at cancellation issues next.

jburel commented:
When the work is done, this page http://help.openmicroscopy.org/getting-started-5.html#adding-data could be updated (e.g. 6 and 7.)

dominikl commented:
I'll remove the 'UpgradeCheck' option again. Also noticed that https://www.openmicroscopy.org/community/viewtopic.php?f=4&t=8560 is only partially fixed, if an import already fails before file upload starts, then the importer still gets into this unusable state and has to be closed/reopenend.

dominikl committed:
More bug fixes
- Fixed the naming of some labels
- Removed UpgradeCheck option
- Fixed issue with Importer hang on failed upload

pwalczysko commented:
For today, I just managed to download the insight and connect to eel (not sure how to connect to east ci ?). Imported just two files, which worked. Will continue tomorrow.

pwalczysko commented:
RFE: ``Skip compute`` sounds just not right to me. Maybe ``Skip computation`` or even ``Import speed-up`` as a tab name and then above the checkboxes, add ``Shorten the import by not computing....``

cc @mtbc  ?

pwalczysko commented:
RFE: See below: this looks to me like a waste of real estate - there is so much space, but I have just three tiny lines squeezed into the middle.

![screen shot 2018-08-07 at 13 20 35](https://user-images.githubusercontent.com/2478303/43775495-d34a0078-9a44-11e8-9394-8955a858b54f.png)

pwalczysko commented:
RFE: give me the number of uploaded and processed files in the final report.
![screen shot 2018-08-07 at 13 23 25](https://user-images.githubusercontent.com/2478303/43775593-260902b4-9a45-11e8-8e9e-a6fc770dc3fb.png)

dominikl commented:
What else should we put there to fill the space? Shrinking the dialog after starting the import would be even more weird.

dominikl committed:
Some more small fixes
- Align elements
- Set spinner only active when import started

pwalczysko commented:
>What else should we put there to fill the space? Shrinking the dialog after starting the import would >be even more weird.

I was thinking the other way, blow up the feature a bit, give more spaces between lines, bigger font, bigger spinners etc.

pwalczysko commented:
Bug: ``Cancel All`` causes a hang.

   - have a folder with 143 images
   - import it
   - after some images are shown as loaded and processed, click ``Cancel All``
   - observe indefinite spinning of the spinners, the counters stop incrementing, this takes more than 15 minutes -> close via Force Quit


Observe that in Insight log, there is
```
2018-08-07 18:38:34,301 WARN  [    ome.formats.OMEROMetadataStoreClient] ( Thread-16) Exception closing 6409aomero.api.MetadataStore -t -e 1.0:tcp -h eel.open,,,py.org -p 43815 -t infinite 
Ice.ObjectNotExistException: null
	at IceInternal.AsyncResultI.__wait(AsyncResultI.java:276) ~[ice.jar:na]
	at omero.api.MetadataStorePrxHelper.end_close(MetadataStorePrxHelper.java:1646) ~[blitz.jar:na]
	at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1537) ~[blitz.jar:na]
	at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1524) ~[blitz.jar:na]
	at ome.formats.OMEROMetadataStoreClient.closeQuietly(OMEROMetadataStoreClient.java:1042) [blitz.jar:na]
	at ome.formats.OMEROMetadataStoreClient.closeServices(OMEROMetadataStoreClient.java:1072) [blitz.jar:na]
	at omero.gateway.Connector.shutdownImports(Connector.java:848) [blitz.jar:na]
	at omero.gateway.Connector.closeImport(Connector.java:716) [blitz.jar:na]
	at omero.gateway.Gateway.closeImport(Gateway.java:527) [blitz.jar:na]
	at org.openmicroscopy.shoola.env.data.OMEROGateway.closeImport(OMEROGateway.java:7181) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importCandidates(OmeroImageServiceImpl.java:257) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1372) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:73) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:48) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:97) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276) [omero.insight.jar:na]
	at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91) [omero.insight.jar:na]
	at java.lang.Thread.run(Thread.java:748) [na:1.8.0_171]
```

Edit: Note that if you import less than 100 files the bug does not occur -> instead, normal ending of the process happens.

jburel committed:
Remove usage of deprecated fields

dominikl committed:
Merge pull request #11 from jburel/lightweight_importer_4

deprecated fields

jburel committed:
Consider the specified container type

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/env/data/OMEROGateway.java:
```diff
@@ -5966,24 +5969,37 @@ Object importImageFile(SecurityContext ctx, ImportableObject object,
+            ImportConfig config = new ImportConfig();
+            config.checksumAlgorithm.set(object.getChecksumAlgorithm());
+            if (object.skipThumbnails()) {
+                config.doThumbnails.set(Boolean.FALSE);
```
With autoboxing it's fine to just write `false` and `true`.

joshmoore commented:
@dominikl: will you have a chance to look into the cancel issue?

dominikl commented:
Yes, I hope I can fix that today or tomorrow. If the rest is ok, the PR should be hopefully finished then.

dominikl committed:
Fixed issue with CancelAll

dominikl committed:
Just use true/false

dominikl committed:
Few UI improvements

dominikl commented:
- The "Cancel All" issue should be fixed now ( fixes https://github.com/openmicroscopy/openmicroscopy/pull/5783#issuecomment-411141746 )
- Instead of "Finished" simply the number of files imported is shown (fixes https://github.com/openmicroscopy/openmicroscopy/pull/5783#issuecomment-411037639 )
- Renamed the title of the "Skip compute" tab to "Import speed-up", but added the text "Shorten the import by skipping..." instead of "Shorten the import by not computing..." (because "not computing" doesn't really go together with "Thumbnail generation") ( fixes https://github.com/openmicroscopy/openmicroscopy/pull/5783#issuecomment-410811938 )

joshmoore commented:
I haven't been able to reproduce it, but while importing 300 images, I navigated away from the GUI after 35 had completed. On navigating back, this number hadn't increased but began increasing as soon as I looked at it. Does this seem at all plausible?

Edit: parallel import was happening on the same server in the background so this should not be considered too terribly serious unless we see it in isolation.

joshmoore commented:
NB: Likely as an additional RFE for the future, knowing the time taken in the new simple UI might help fill some space.

dominikl committed:
Count cancelled status correctly and take errors into account

dominikl committed:
Fix NPE and avoid double UI update

dominikl commented:
Had to fix another issue: If a directory with 200+ files contains broken files, the spinners in the UI didn't stop after cancelling because these broken imports weren't counted as "finished".

When it comes to mixing single images and directories for import the UI is still a bit strange. For example if you select a single image to import and then also a directory (e.g. with 250 files), the initial total number of imports will be 2 and then jumps to 251 when the directory import starts. Also you can't cancel the directory import then. That's because the directory import is then treated as a "unit" and you can only cancel an import before it starts. This is just how the importer works...

There's definitely a lot space for further improvements, although in my opinion a complete reimplementation would then probably make more sense.

rgozim commented:
Tested import "cancel all" UI. 

Before cancel:
![screen shot 2018-08-23 at 15 30 29](https://user-images.githubusercontent.com/3717090/44531847-a6c02080-a6e9-11e8-9655-13059aa1f7a2.png)

After cancel:
![screen shot 2018-08-23 at 15 30 53](https://user-images.githubusercontent.com/3717090/44531861-aaec3e00-a6e9-11e8-88eb-769baf73914a.png)

The only thing that maybe needs tweaked, is the loading spinner still present in the top tab.

jburel commented:
The "Tick/Red cross" visually implies that some files failed to import.
It does not really indicate that the import was cancelled

dominikl commented:
Ok, I change it to the normal green tick icon. Unless someone has an idea for a better one indicating the "cancelled" state.

jburel commented:
Something is a bit misleading in the UI since processed should only indicate the ones that have been uploaded and technically processed so in that case 19/19.
If a user cancel you could add an entry "cancelled x" before Errors

dominikl committed:
Show number of cancelled imports

dominikl commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/view/ImporterUIElementLight.java:
```diff
@@ -164,10 +167,24 @@ private void buildGUI() {
+        c.gridy = 3;
+        c.fill = GridBagConstraints.HORIZONTAL;
+        c.gridwidth = 2;
+        c.anchor = GridBagConstraints.WEST;
```
Right-justifies the labels (addresses Marks [comment](https://github.com/openmicroscopy/openmicroscopy/pull/5783/commits/144893ef5090a2c443c10fa95d9f859a4081bf0a#r206153748) )

dominikl commented:
With the last commit the number of cancelled imports is shown (only displayed if there are cancelled imports), and the "Processed" progressbar is now tracking the number of processed files in relation to the number of uploaded files (not in relation to the total files to import).

dominikl committed:
Fixed another issue with counts; added 'scanning' status display

dominikl commented:
Just realized while preparing the changes for the import testing scenario that the "Scanning" phase when bioformats checks the local files can take a bit of time when importing non-trivial images; therefore added another row displaying a "Scanning" label and spinner to indicate that something's happening.

![screen shot 2018-08-29 at 14 48 47](https://user-images.githubusercontent.com/6575139/44792207-432b6c80-ab9b-11e8-8424-749006560093.png)

mtbc commented on components/insight/SRC/org/openmicroscopy/shoola/agents/fsimporter/actions/CancelAction.java:
```diff
@@ -25,10 +25,13 @@
+
 import javax.swing.Action;
 
 //Third-party libraries
```
fine to just get rid of this middle bit

joshmoore commented:
Hmmm.... I still have a green check and a spinning wheel after "cancel all". I checked with `jstack`; no threads are actively working, but this is logged:

```
4:11,479 INFO  [      o.f.i.transfers.UploadFileTransfer] ( Thread-41) Transferring /tmp/fakes/image_003.fake...
2018-08-30 13:14:13,494 WARN  [    ome.formats.OMEROMetadataStoreClient] ( Thread-41) Exception closing 1dcc725f-ba5e-4535-a949-c246bbc26774/7a44c67d-bc26-457c-8dfe-ba93857f237eomero.api.MetadataStore -t -e 1.0:tcp -h eel.openmicroscopy.org -p 37563 -t infinite
Ice.ObjectNotExistException: null
        at IceInternal.AsyncResultI.__wait(AsyncResultI.java:276) ~[ice.jar:na]
        at omero.api.MetadataStorePrxHelper.end_close(MetadataStorePrxHelper.java:1646) ~[blitz.jar:na]
        at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1537) ~[blitz.jar:na]
        at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1524) ~[blitz.jar:na]
        at ome.formats.OMEROMetadataStoreClient.closeQuietly(OMEROMetadataStoreClient.java:1042) [blitz.jar:na]
        at ome.formats.OMEROMetadataStoreClient.closeServices(OMEROMetadataStoreClient.java:1072) [blitz.jar:na]
        at omero.gateway.Connector.shutdownImports(Connector.java:848) [blitz.jar:na]
        at omero.gateway.Connector.closeImport(Connector.java:716) [blitz.jar:na]
        at omero.gateway.Gateway.closeImport(Gateway.java:527) [blitz.jar:na]
        at org.openmicroscopy.shoola.env.data.OMEROGateway.closeImport(OMEROGateway.java:7183) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1013) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:73) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:48) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:97) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91) [omero.insight.jar:na]
        at java.lang.Thread.run(Thread.java:748) [na:1.8.0_181]
2018-08-30 13:14:13,888 WARN  [    ome.formats.OMEROMetadataStoreClient] ( Thread-41) Exception closing 1dcc725f-ba5e-4535-a949-c246bbc26774/7a44c67d-bc26-457c-8dfe-ba93857f237eomero.api.MetadataStore -t -e 1.0:tcp -h eel.openmicroscopy.org -p 37563 -t infinite
Ice.ObjectNotExistException: null
        at IceInternal.AsyncResultI.__wait(AsyncResultI.java:276) ~[ice.jar:na]
        at omero.api.MetadataStorePrxHelper.end_close(MetadataStorePrxHelper.java:1646) ~[blitz.jar:na]
        at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1537) ~[blitz.jar:na]
        at omero.api.MetadataStorePrxHelper.close(MetadataStorePrxHelper.java:1524) ~[blitz.jar:na]
        at ome.formats.OMEROMetadataStoreClient.closeQuietly(OMEROMetadataStoreClient.java:1042) [blitz.jar:na]
        at ome.formats.OMEROMetadataStoreClient.closeServices(OMEROMetadataStoreClient.java:1072) [blitz.jar:na]
        at omero.gateway.Connector.shutdownImports(Connector.java:848) [blitz.jar:na]
        at omero.gateway.Connector.closeImport(Connector.java:716) [blitz.jar:na]
        at omero.gateway.Gateway.closeImport(Gateway.java:527) [blitz.jar:na]
        at org.openmicroscopy.shoola.env.data.OMEROGateway.closeImport(OMEROGateway.java:7183) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.OmeroImageServiceImpl.importFile(OmeroImageServiceImpl.java:1013) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.importFile(ImagesImporter.java:73) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter.access$000(ImagesImporter.java:48) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.calls.ImagesImporter$1.doCall(ImagesImporter.java:97) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.BatchCall.doStep(BatchCall.java:144) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.CompositeTask.doStep(CompositeTask.java:226) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.env.data.views.CompositeBatchCall.doStep(CompositeBatchCall.java:126) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.exec(ExecCommand.java:165) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.ExecCommand.run(ExecCommand.java:276) [omero.insight.jar:na]
        at org.openmicroscopy.shoola.util.concur.tasks.AsyncProcessor$Runner.run(AsyncProcessor.java:91) [omero.insight.jar:na]
```

dominikl commented:
Can't replicate. Is this happening after the import runs for a longer time?

joshmoore commented:
On a slowish connection (SSH tunnel to eel from home) after 3 or so imports from 300 have succeeded.

dominikl committed:
Take component itself into account if no children

dominikl committed:
Remove unneccessary line break

dominikl commented:
Fixed another potential issue, but I don't think that was the reason for the `ObjectNotExistException` exception, which I still can replicate.

joshmoore commented:
Hmmm.... :+1: no spinning wheel this time:
<img width="366" alt="screen shot 2018-08-31 at 13 23 31" src="https://user-images.githubusercontent.com/88113/44909820-1acd7a80-ad21-11e8-8ab5-85f43531aaca.png">

I *still* have the exceptions (see below).

```
2018-08-31 13:23:29,567 WARN  [    ome.formats.OMEROMetadataStoreClient] ( Thread-13) Exception closing bd98a4fc-e908-42fc-b88d-729dd6447085/85da98d0-36cb-47e8-8d04-6057797a9c64omer
o.api.MetadataStore -t -e 1.0:tcp -h eel.openmicroscopy.org -p 36372 -t infinite
Ice.ObjectNotExistException: null
```

but testing on develop:

```
2018-08-31 13:29:33,817 WARN  [    ome.formats.OMEROMetadataStoreClient] ( Thread-14) Exception closing 18610c5a-eb28-4cd1-8b06-03a254bd4e7c/d8c6729f-cb14-4722-8a89-e953d73c8e54omero.api.MetadataStore -t -e 1.0:tcp -h eel.openmicroscopy.org -p 36372 -t infinite
Ice.ObjectNotExistException: null
```

*Not* a regression.

dominikl commented:
That's great! I'll do another run through of the import testing scenario to check that there's nothing broken of the normal (less than 200 files) import workflow. But can do that as part of the RC testing too, if necessary.

joshmoore commented:
Let's call it a day/week/month/milestone and merge. Fixes can come on top of the RC. Thanks to @dominikl @rgozim etal.
